pgjdbc/src/main/java/org/postgresql/PGConnection.java

```diff
@@ -244,6 +251,41 @@ public interface PGConnection {
    */
   PGReplicationConnection getReplicationAPI();
 
+  /**
+   * Change a user's password to the specified new password.
+   *
+   * <p>
+   * If the specific encryption type is not specified, this method defaults to querying the database server for the server's default password_encryption.
+   * This method does not send the new password in plain text to the server.
+   * Instead, it encrypts the password locally and sends the encoded hash so that the plain text password is never sent on the wire.
+   * </p>
+   *
+   * <p>
+   * Acceptable values for encryptionType are null, "md5", or "scram-sha-256".
+   * Users should avoid "md5" unless they are explicitly targeting an older server that does not support the more secure SCRAM.
+   * </p>
+   *
+   * @param user The username of the database user
+   * @param passwordSource Callback to supply the plain text of the user's password. The char[] return value of this function will be destroyed after usage.
+   * @param encryptionType The type of password encryption to use or null if the database server default should be used.
+   * @throws SQLException If the password could not be altered
+   */
+  default void alterUserPassword(String user, Supplier<char[]> passwordSource, @Nullable String encryptionType) throws SQLException {
```

Comment:
What is the purpose of `Supplier` in the signature?
I think a regular `char[]` should be enough.

Reply:
The rationale is that by providing a callback rather than a value, it's explicit that the ownership of the char[] object lies with the method and not the caller. This also matches how we provide passwords in the auth callbacks.

Reply:
A simple use case would match the tests where a string is simply converted to a char[].

A more elaborate use case could have the callback itself read from a KMS. It's the ultimate in flexibility while ensuring the char[] never lingers.

Reply:
You don't need to have supplier for that. Just document that in the javadoc: `password will be erased after use` and that is it. Supplier makes it harder for the clients to pass the value.

Reply:
If the caller is supplying a char[] then the caller owns that object and the caller must wipe it.

If we want to wipe it within the method, we need to own the char[] and the way to signify that is to have it's creation happen via the supplier. Plus we get the flexibility for the more elaborate use cases.

Plus it's actually easier with the supplier style as the only usage within the driver of the password is wrapped around the supplier which ensures that the raw char[] value never lingers.

If the method signature has the `char[]` and overload A calls overload B (which is supposed to eventually wipe the password), how do you ensure the password was wiped and something else did not throw an exception before that step?

Reply:
No. We declare in javadoc who owns the object and that is it.

>how do you ensure the password was wiped and something else did not throw an exception before that step?

Wipe the password in `finally`

Reply:
That's my point. You end up with the same code in every overload as otherwise you never know if a change to the overload impacts when exactly the password gets cleared.

Anyway, I'm not keen to change it. The callback method is the most versatile and covers all the more complex use cases I've described.

If we're deadlocked on this, then get someone else to agree to the change in the signature. Otherwise I'm done with this and plan to rebase and merge.

Reply:
> You end up with the same code in every overload

There are only a few methods dealing with passwords, and, yes, security code is not the most easy to implement correctly.

Neither https://docs.oracle.com/javase%2Ftutorial%2Fuiswing%2F%2F/components/passwordfield.html nor https://docs.oracle.com/javase/8/docs/technotes/guides/security/crypto/CryptoSpec.html suggest using `Supplier` in the method parameters.